File: legacy/src/utils/input-manager.ts
/**
 * Input Manager — Centralized input ownership system.
 *
 * When a game or overlay takes focus, it calls setInputOwner("game-id").
 * While an owner is set:
 *   - The main.js click handler does NOT delegate to room.handleClick()
 *   - The main.js render loop does NOT call room.animate()
 *   - Only the owning game's own listeners receive events
 *
 * On game exit, the game calls clearInputOwner() to release control.
 *
 * Usage:
 *   import { setInputOwner, clearInputOwner } from "../utils/input-manager.js";
 *   // On game launch:
 *   setInputOwner("cosmic-pool");
 *   // On game exit:
 *   clearInputOwner();
 */

let _currentOwner: string | null = null;
let _mainCanvas: HTMLElement | null = null;

/**
 * Set the current input owner. While owned, the main room
 * click/hover/animate pipeline is fully suppressed.
 *
 * @param {string} gameId — unique identifier for the owning game
 */
export function setInputOwner(gameId: string): void {
  _currentOwner = gameId;
  console.log(`🎮 Input owner set: ${gameId}`);

  // Disable pointer events on the main Three.js canvas so
  // clicks fall through to game overlays only
  if (!_mainCanvas) {
    _mainCanvas = document.getElementById("room-canvas");
  }
  if (_mainCanvas) {
    _mainCanvas.style.pointerEvents = "none";
  }
}

/**
 * Clear input ownership — restores normal room interaction.
 */
export function clearInputOwner(): void {
  const prev = _currentOwner;
  _currentOwner = null;
  console.log(`🎮 Input owner cleared (was: ${prev})`);

  // Restore pointer events on the main canvas
  if (_mainCanvas) {
    _mainCanvas.style.pointerEvents = "auto";
  }
}

/**
 * Get the current input owner ID (or null).
 * @returns {string|null}
 */
export function getInputOwner(): string | null {
  return _currentOwner;
}

/**
 * Check if any game currently owns the input.
 * @returns {boolean}
 */
export function isInputOwned(): boolean {
  return _currentOwner !== null;
}

File: crates/pk-core/src/input/input_manager.rs
//! Centralized Input Ownership Manager — Focus arbitrator suppressing background room interactions.
//!
//! PORTS: `legacy/src/utils/input-manager.ts`

use std::sync::Mutex;

static INPUT_OWNER: Mutex<Option<String>> = Mutex::new(None);

pub fn set_input_owner(game_id: &str) {
    if let Ok(mut lock) = INPUT_OWNER.lock() {
        *lock = Some(game_id.to_string());
    }
}

pub fn clear_input_owner() {
    if let Ok(mut lock) = INPUT_OWNER.lock() {
        *lock = None;
    }
}

pub fn get_input_owner() -> Option<String> {
    if let Ok(lock) = INPUT_OWNER.lock() {
        lock.clone()
    } else {
        None
    }
}

pub fn is_input_owned() -> bool {
    if let Ok(lock) = INPUT_OWNER.lock() {
        lock.is_some()
    } else {
        false
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct InputOwnerState {
    pub current_owner: Option<String>,
}

impl InputOwnerState {
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the current input owner. While owned, background interaction is suppressed.
    pub fn set_owner(&mut self, owner: &str) {
        self.current_owner = Some(owner.to_string());
    }

    /// Clear input ownership — restores normal room interaction.
    pub fn clear_owner(&mut self) {
        self.current_owner = None;
    }

    /// Check if any game or overlay currently owns the input.
    pub fn is_owned(&self) -> bool {
        self.current_owner.is_some()
    }

    /// Get the current input owner ID (or None).
    pub fn get_owner(&self) -> Option<&str> {
        self.current_owner.as_deref()
    }
}
